add support for MONGODB_EXTRA_DATABASES

efrecon opened the pull request:
**Description of the change**

This adds support for giving the user at MONGODB_USERNAME access to more
than one database through recognising a comma-separated list of
databases in the (new) MONGODB_EXTRA_DATABASES environment variable.

**Benefits**

The new variable allows the creation of several databases on the first run in a backward compatible manner.

**Possible drawbacks**

Having two possible variables where to express the databases to be created the first time a container is started is not always intuitive. But it guarantees that existing deployments will not break at upgrade time.

**Applicable issues**

<!-- Enter any applicable Issues here (You can reference an issue using #) -->

**Additional information**

The need for this PR originates from an [issue](https://github.com/Countly/countly-server/issues/1937) in the [countly](https://count.ly/) server.

juan131 reviewed (commented):
Thanks so much for this great contribution! We really appreciate it!

We had similar requests like this one in the past in database solutions where users want to create N users or N databases during the container startup.

We came with a different approach in assets such as [Kafka](https://github.com/bitnami/bitnami-docker-kafka). Instead of adding a new env. variable, we deprecated `XXX_USER` and `XXX_PASSWORD` (expecting a string with a single user and password, respectively) in favor of `XXX_USERS` and `XXX_PASSWORDS` (expecting a comma-separated list of users and passwords, respectively)

- https://github.com/bitnami/bitnami-docker-kafka#interbroker-communications
- https://github.com/bitnami/bitnami-docker-kafka#250-debian-10-r111

I would go for this approach transforming `MONGODB_DATABASE` into `MONGODB_DATABASES`. This way, we'll expect `MONGODB_USERS`, `MONGODB_PASSWORDS`, and `MONGODB_DATABASES` to be comma-separated lists with the **exact same length** and we'll take the "i" element of each list to create a user/password/database combi. 
We can simply skip creating a user when it's provided twice in the list.

The main benefit:

- You can have multiple databases, but also having different users/passwords on each database.

What do you think?

efrecon commented:
I think that your approach is indeed better. The only drawback is that it prevents the use of `,` and `;` signs in the passwords, as I [understand](https://github.com/bitnami/bitnami-docker-kafka/blob/ebab4729d2faff06e9c695e3a942b38d85c9fb61/2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh#L212) it. But this is probably a minor issue when using automatically generated (and random) passwords as generation rules generally are flexible enough.

I saw that the kafka implementation already has support for [aliasing](https://github.com/bitnami/bitnami-docker-kafka/blob/ebab4729d2faff06e9c695e3a942b38d85c9fb61/2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh#L84) variables, which facilitate migration of existing setups to the new variables with an `S` at the end.

I will have a go at modifying my PR in the coming days.

juan131 commented:
> The only drawback is that it prevents the use of `,` and `;` signs in the passwords, as I understand it. But this is probably a minor issue when using automatically generated (and random) passwords as generation rules generally are flexible enough.

Good point! That's a drawback... We can also change the approach and only accepts spaces as separator for the passwords.

> I will have a go at modifying my PR in the coming days.

Thanks so much

efrecon commented:
I have now updated my PR to recognize variables ending with an `S` as @juan131 suggested. The implementation uses the `,` or `;` as separators. I believe this is better as these are the characters that seem to be used in other Bitnami images for the same kind of purpose. I have updated the README to point at the difference and that the new implementation might break a few edge cases. I hope that my changes to the README properly reflect the coming release tags, please modify as appropriate.

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -13,6 +13,85 @@
 . /opt/bitnami/scripts/libfs.sh
 . /opt/bitnami/scripts/libnet.sh
 
+########################
+# Create alias for environment variable, so both can be used
+# Globals:
+#   None
+# Arguments:
+#   $1 - Alias environment variable name
+#   $2 - Original environment variable name
+# Returns:
+#   None
+#########################
+mongodb_declare_alias_env() {
+    local -r alias="${1:?missing environment variable alias}"
+    local -r original="${2:?missing original environment variable}"
+    if printenv "${original}" > /dev/null; then
+        export "$alias"="${!original:-}"
+    fi
+}
+
+
+########################
+# Map MongoDB legacy environment variables to the new names
+# Globals:
+#   MONGODB_*
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+mongodb_create_alias_environment_variables() {
+    mongodb_declare_alias_env "MONGODB_USERNAMES" "MONGODB_USERNAME"
+    mongodb_declare_alias_env "MONGODB_PASSWORDS" "MONGODB_PASSWORD"
+    mongodb_declare_alias_env "MONGODB_DATABASES" "MONGODB_DATABASE"
+    mongodb_declare_alias_env "ALLOW_EMPTY_PASSWORDS" "ALLOW_EMPTY_PASSWORD"
+}
```
This shouldn't be necessary. Instead, let's solve the aliases in the `mongodb-env.sh`

efrecon replied:
I agree that moving this to `mongodb-env.sh` is simpler. However, this is aligned on what other (similar) bitnami's repos are doing. For example,
+ The kafka implementation uses aliasing starting from [this](https://github.com/bitnami/bitnami-docker-kafka/blob/9d92b711132bb42e133f6d88e124301070c2fd12/2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh#L127) point an onwards in the next proc.
+ The consul implementation uses the same implementation [here](https://github.com/bitnami/bitnami-docker-consul/blob/690f0c155b0621375d68fabab90fed21419aa2df/1/debian-10/rootfs/opt/bitnami/scripts/libconsul.sh#L13) and onwards.

I *think* that keeping the implementation similar and align across all of Bitnami's repositories is a benefit and should be favoured. Opinion?

juan131 replied:
I agree with aligning with other containers do. These `xxx-env.sh` files are kind of new in our "standard" and we're trying to move all our apps to use them.

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/mongodb-env.sh:
```diff
@@ -39,6 +39,7 @@ mongodb_env_vars=(
     MONGODB_USERNAME
     MONGODB_PASSWORD
     MONGODB_DATABASE
+    MONGODB_EXTRA_DATABASES
```
You can drop this. Instead let's declare the new env. vars here (keeping the old ones for backwards compatibility):
```suggestion
    MONGODB_USERNAMES
    MONGODB_PASSWORDS
    MONGODB_DATABASES
```

efrecon replied:
Good catch for the `MONGODB_EXTRA_DATABASES` removal!

As for keeping the old variables there, I ... don't think that they should be there if we opt to align ourselves on other Bitnami's images implementations. For example, the old variable `KAFKA_SEGMENT_BYTES` does not appear [here](https://github.com/bitnami/bitnami-docker-kafka/blob/master/2/debian-10/rootfs/opt/bitnami/scripts/kafka-env.sh) but is still aliased and referenced from [here](https://github.com/bitnami/bitnami-docker-kafka/blob/9d92b711132bb42e133f6d88e124301070c2fd12/2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh#L187)

juan131 replied:
As I mentioned in my previous comment, we're standardizing to use this file to manage everything related to the declaration of the env. variables and assigning its default values.

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/mongodb-env.sh:
```diff
@@ -112,10 +113,10 @@ export MONGODB_DEFAULT_SYSTEM_LOG_VERBOSITY="0"
+export MONGODB_USERNAMES="${MONGODB_USERNAMES:-}"
+export MONGODB_PASSWORDS="${MONGODB_PASSWORDS:-}"
+export MONGODB_DATABASES="${MONGODB_DATABASES:-}"
+export ALLOW_EMPTY_PASSWORDS="${ALLOW_EMPTY_PASSWORDS:-no}"
```
I wouldn't change this env. variable. Why do you need a list? I mean, can't we simply assume that `ALLOW_EMPTY_PASSWORD` allows setting empty passwords for every user?

efrecon replied:
It was just to keep the trailing `S` consistent. But maybe you are right!

juan131 replied:
Yes but please note there's a `MONGODB_ROOT_PASSWORD` env. var affected by `ALLOW_EMPTY_PASSWORD`. We should definitely keep it as it is and assume this boolean affect every single password.

efrecon replied:
Yes, as far as I remember, my modification pays attention to this relation (albeit using `MONGODB_ROOT_PASSWORDS` right now, but that we agreed to change back to `MONGODB_ROOT_PASSWORD` )

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/mongodb-env.sh:
```diff
@@ -112,10 +113,10 @@ export MONGODB_DEFAULT_SYSTEM_LOG_VERBOSITY="0"
-export MONGODB_PASSWORD="${MONGODB_PASSWORD:-}"
-export MONGODB_DATABASE="${MONGODB_DATABASE:-}"
-export ALLOW_EMPTY_PASSWORD="${ALLOW_EMPTY_PASSWORD:-no}"
+export MONGODB_USERNAMES="${MONGODB_USERNAMES:-}"
+export MONGODB_PASSWORDS="${MONGODB_PASSWORDS:-}"
+export MONGODB_DATABASES="${MONGODB_DATABASES:-}"
```
We can manage the aliases here:
```suggestion
export MONGODB_USERNAMES="${MONGODB_USERNAMES:-"${MONGODB_USERNAME:-}"}"
export MONGODB_PASSWORDS="${MONGODB_PASSWORDS:-"${MONGODB_PASSWORD-}"}"
export MONGODB_DATABASES="${MONGODB_DATABASES:-"${MONGODB_DATABASE:-}"}"
```

efrecon replied:
See my comment around "procedural aliasing" above

juan131 replied:
Replied

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -82,11 +162,38 @@ Available options are 'primary/secondary/arbiter/hidden'"
-        error_message="The MONGODB_PASSWORD environment variable is empty or not set. Set the environment variable ALLOW_EMPTY_PASSWORD=yes to allow the container to be started with blank passwords. This is only recommended for development."
-        print_validation_error "$error_message"
+    # Deprecation warnings
+    if [ -n "${MONGODB_USERNAME-}" ]; then
+        warn "The use of MONGODB_USERNAME is deprecated, use MONGODB_USERNAMES instead. ',' and ';' characters are now reserved for separating fields, fields cannot use any default shell delimiters."
+    fi
+    if [ -n "${MONGODB_PASSWORD-}" ]; then
+        warn "The use of MONGODB_PASSWORD is deprecated, use MONGODB_PASSWORDS instead. ',' and ';' characters are now reserved for separating fields, fields cannot use any default shell delimiters."
+    fi
+    if [ -n "${MONGODB_DATABASE-}" ]; then
+        warn "The use of MONGODB_DATABASE is deprecated, use MONGODB_DATABASES instead. ',' and ';' characters are now reserved for separating fields, fields cannot use any default shell delimiters."
+    fi
```
```suggestion
    if [[ -n "$MONGODB_USERNAME" ]]; then
        warn "The use of MONGODB_USERNAME is deprecated, use MONGODB_USERNAMES instead. Also, note that ',' and ';' characters are now reserved for separating fields, fields cannot use any default shell delimiters."
    fi
    if [[ -n "$MONGODB_PASSWORD" ]]; then
        warn "The use of MONGODB_PASSWORD is deprecated, use MONGODB_PASSWORDS instead. Also, note that ',' and ';' characters are now reserved for separating fields, fields cannot use any default shell delimiters."
    fi
    if [[ -n "$MONGODB_DATABASE" ]]; then
        warn "The use of MONGODB_DATABASE is deprecated, use MONGODB_DATABASES instead. Also, note that ',' and ';' characters are now reserved for separating fields, fields cannot use any default shell delimiters."
    fi
```

efrecon replied:
As for the `[[` and `]]` construct in the if-statements, I don't think that it is needed in that case, but I agree that it complies to the remaining coding style of the script.

When it comes to the `${XXX-}}` constructs, these were intentional to let the code work even though the variables were not even declared.

juan131 replied:
The `[[ ]]` is just a matter of conventions and follow the same style we use in the rest of the files. Regarding using `${XXX-}}`, that shouldn't be necessary if the variables are declared at `mongodb-env.sh` since you're ensuring they will exist.

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -82,11 +162,38 @@ Available options are 'primary/secondary/arbiter/hidden'"
+        warn "The use of MONGODB_DATABASE is deprecated, use MONGODB_DATABASES instead. ',' and ';' characters are now reserved for separating fields, fields cannot use any default shell delimiters."
+    fi
+
+    if [[ -n "$MONGODB_DATABASES" ]]; then
+        mongodb_auth
+        # Verify there as many usernames as passwords as databases
+        if [[ "${#usernames[@]}" -ne "${#databases[@]}" ]]; then
+            print_validation_error "Specify the same number of users on MONGODB_USERNAMES as the number of databases on MONGODB_DATABASES!"
+        fi
+        if [[ "${#usernames[@]}" -ne "${#passwords[@]}" ]]; then
+            print_validation_error "Specify the same number of passwords on MONGODB_PASSWORDS as the number of users on MONGODB_USERNAMES!"
+        fi
+    fi
```
So... This makes mandatory to set  `MONGODB_USERNAMES` and `MONGODB_PASSWORDS` when `MONGODB_DATABASES` is set. We should mention that in the **README.md**

efrecon replied:
The README says that there should be as many datbases as there are users and passwords. Where would you suggest that we make this clearer?

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -82,11 +162,38 @@ Available options are 'primary/secondary/arbiter/hidden'"
+    # Verify empty passwords
+    if is_boolean_yes "$ALLOW_EMPTY_PASSWORDS"; then
+        warn "You set the environment variable ALLOW_EMPTY_PASSWORDS=${ALLOW_EMPTY_PASSWORDS}. For safety reasons, do not use this flag in a production environment."
+    elif [[ -n "$MONGODB_DATABASES" ]]; then
```
Why this check? Shouldn't I be able to pass a list of users and passwords to use without passing a list of databases?

efrecon replied:
See my comment above. If you want several users for the same database, you will have to repeat the name of the database several times, with `,`-signs in between. This is the only way to make the code deterministic in all cases as it loops over the databases, users and passwords at the same time.

juan131 replied:
But we should make possible to create a list o users with no associated databases, right?

efrecon replied:
That ... wasn't part of the original implementation. But that would be a good addition. I will have to treat this as a special case, but this can easily be done I think. I will look into this.

juan131 replied:
Thanks so much!

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -13,6 +13,85 @@
+#   None
+#########################
+mongodb_auth() {
+    IFS="$(mongodb_field_separator "${MONGODB_DATABASES}")" read -r -a databases <<< "${MONGODB_DATABASES}"
+    IFS="$(mongodb_field_separator "${MONGODB_USERNAMES:-}")" read -r -a usernames <<< "${MONGODB_USERNAMES:-}"
+    IFS="$(mongodb_field_separator "${MONGODB_PASSWORDS:-}")" read -r -a passwords <<< "${MONGODB_PASSWORDS:-}"
```
Shouldn't this be enough?
```suggestion
read -r -a databases <<< "$(tr ',;' ' ' <<< "${MONGODB_DATABASES}")"
read -r -a users <<< "$(tr ',;' ' ' <<< "${MONGODB_USERNAMES}")"
read -r -a passwords <<< "$(tr ',;' ' ' <<< "${MONGODB_PASSWORDS}")"
```

efrecon replied:
No, I don't think so.
+ I believe that your implementation would prevent the use of space, tabs and newlines in the passwords, as these are the regular delimiters in shell. (of those maybe space is the one to look for).
+ Also the use of `${xx:-}` was conscious to permit reading from empty and non-existing variables.

juan131 replied:
The `${xx:-}` shouldn't be necessary if we declare the env. vars at `mongodb-env.sh`. Regarding the allowed delimiters you're right, my suggestion will treat spaces, tabs, newlines, commas and semicolons as delimiters.
My question is: does MongoDB accepts spaces, tabs and newlines in the passwords?

efrecon replied:
Yes. MongoDB accepts spaces in the passwords. The documentation does not [mention](https://docs.mongodb.com/manual/reference/method/db.updateUser/) any restrictions for passwords and I verified through creating a Mongo instance with a root password that contained a space. I would suggest that we keep my version of the code.

juan131 replied:
All right! Let's use whatever approach that is less restrictive with what characters can be included in the passwords.

juan131 reviewed (changes requested):
Thanks so much for implementing my suggestion!! I really appreciate it.

Find some comments below to simplify the logic.

efrecon commented:
@juan131 I *think* that it was simpler to push yet a new change incorporating the results of our discussion above. I hope you don't mind, as this might give you a bit of extra work.

juan131 reviewed (commented):
Thanks so much for the changes @efrecon !! I'll give it a try during the day to the different combinations and share my feedback to you.

efrecon commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -13,6 +13,50 @@
+# Returns:
+#   The separator used within that list
+#########################
+mongodb_field_separator() {
+    if printf %s\\n "$1" | grep -q ','; then
+        printf %s\\n ','
+    elif printf %s\\n "$1" | grep -q ';'; then
+        printf %s\\n ';'
+    fi
```
`echo` automatically adds a newline anyway. You have to explicitly use the `-n` [option](https://linux.die.net/man/1/echo) to avoid the newline. I tend to avoid `echo` because it has side-effects with backslash sequences (although this is completely irrelevant here, just a matter of habit). We can use `echo` if it better suits bitnami's scripting guidelines (where are they?)

efrecon commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -30,6 +74,7 @@ mongodb_validate() {
 need to provide the MONGODB_REPLICA_SET_KEY on every node, specify MONGODB_ROOT_PASSWORD \
 in the primary node and MONGODB_INITIAL_PRIMARY_ROOT_PASSWORD in the rest of nodes"
     local error_code=0
+    local usernames databases passwords
```
This is again a matter of opinion, but ... globals are evil. I *think* that making the extraction code visible is preferable. Making the variables global is hardly an optimisation anyway, as extraction happens only twice: once to verify the options given to the script (the env. vars), and a second time to create the users.

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -82,11 +127,47 @@ Available options are 'primary/secondary/arbiter/hidden'"
+        # When the list of database is empty, then all users will be added to
+        # default database.
+        if [[ -z "$MONGODB_DATABASES" ]]; then
+            info "All users specified in MONGODB_USERNAMES will be added to the default database"
```
Also: it was hard to me to find out the default database was "test".

We should include the name of this default database in this warn message.

efrecon replied:
@juan131 do you know where the `test` name is coming from? Is this a default setting in MongoDB itself?

efrecon replied:
I did some more digging and found that the official tutorial [mentions](https://docs.mongodb.com/manual/tutorial/getting-started/#switch-database) the `test` database.

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -503,14 +588,45 @@ EOF
+
+                    result=$(
+                        mongodb_execute 'root' "$MONGODB_ROOT_PASSWORD" "" "127.0.0.1" <<EOF
+db.getSiblingDB(db.stats().db).createUser({ user: '${usernames[i]}', pwd: '${passwords[i]}', roles: [{role: 'readWrite', db: db.getSiblingDB(db.stats().db).stats().db }] })
```
Wouldn't this also work?
```suggestion
db.getSiblingDB(db.stats().db).createUser({ user: '${usernames[i]}', pwd: '${passwords[i]}', roles: [{role: 'readWrite', db: db.stats().db }] })
```

efrecon replied:
I ... don't know. All the code seems to be accessing via `getSiblingDB()`, I simply do not know why. Your proposal seems to work, but maybe are there edge cases?

juan131 reviewed (changes requested):
@efrecon I've trying different combinations and most of them work smoothly. However, i found sth odd trying this:

```yaml
version: '2'

services:
  mongodb:
    image: docker.io/bitnami/mongodb:4.4
    ports:
      - "27017:27017"
    volumes:
      - 'mongodb_data:/bitnami/mongodb'
    environment:
      - ALLOW_EMPTY_PASSWORD=true
      - MONGODB_ROOT_PASSWORD=password123
      - MONGODB_USERNAMES=my_user1,my_user2
      - MONGODB_PASSWORDS=password123
      - MONGODB_DATABASES=my_database1;my_database2

volumes:
  mongodb_data:
    driver: local
```

Instead of creating a second database `my_database` for user `my_user2` with no authentication, it simply skips the creation. See the logs:

```console
mongodb_1  | mongodb 06:59:12.78 INFO  ==> Creating root user...
mongodb_1  | mongodb 06:59:12.89 INFO  ==> Creating user 'my_user1' in database 'my_database1'...
mongodb_1  | mongodb 06:59:12.99 INFO  ==> Users created
```

efrecon commented on README.md:
```diff
@@ -315,26 +315,26 @@ services:
+## Creating users and databases on first run
 
-You can create a user with restricted access to a database while starting the container for the first time. To do this, provide the `MONGODB_USERNAME`, `MONGODB_PASSWORD` and `MONGODB_DATABASE` environment variables.
+You can create users with restricted access to databases while starting the container for the first time. To do this, provide the `MONGODB_USERNAMES`, `MONGODB_PASSWORDS` and `MONGODB_DATABASES` environment variables. In these variables, the characters `,` and `;` are used as field separators. If you do not provide database names in `MONGODB_USERDATABASES`, all users will be created in the defaul ...
```
It is actually neither `MONGODB_USERDATABASES` nor `MONGODB_USER_DATABASES`, but `MONGODB_DATABASES` and nothing else! I'll fix.

efrecon commented:
> @efrecon I've trying different combinations and most of them work smoothly. However, i found sth odd trying this:
> 
> ```yaml
> version: '2'
> 
> services:
>   mongodb:
>     image: docker.io/bitnami/mongodb:4.4
>     ports:
>       - "27017:27017"
>     volumes:
>       - 'mongodb_data:/bitnami/mongodb'
>     environment:
>       - ALLOW_EMPTY_PASSWORD=true
>       - MONGODB_ROOT_PASSWORD=password123
>       - MONGODB_USERNAMES=my_user1,my_user2
>       - MONGODB_PASSWORDS=password123
>       - MONGODB_DATABASES=my_database1;my_database2
> 
> volumes:
>   mongodb_data:
>     driver: local
> ```
> 
> Instead of creating a second database `my_database` for user `my_user2` with no authentication, it simply skips the creation. See the logs:
> 
> ```
> mongodb_1  | mongodb 06:59:12.78 INFO  ==> Creating root user...
> mongodb_1  | mongodb 06:59:12.89 INFO  ==> Creating user 'my_user1' in database 'my_database1'...
> mongodb_1  | mongodb 06:59:12.99 INFO  ==> Users created
> ```

This is because MongoDB itself refuses to create users with blank passwords. So, there is an [if-statement in my code](https://github.com/efrecon/bitnami-docker-mongodb/blob/207810ca0b537a04ceec51fefa67c7cfa0055215/4.4/debian-10/rootfs/opt/bitnami/scripts/libmongodb.sh#L603) that protects around this case. This protection actually existed in the initial code, which uses `&& [[ -n "$MONGODB_PASSWORD" ]]` in the equivalent [test](https://github.com/bitnami/bitnami-docker-mongodb/blob/13b5dd8712c56da2489e37f589e79ec6b4624879/4.4/debian-10/rootfs/opt/bitnami/scripts/libmongodb.sh#L514).

For this reason, the meaning of `ALLOW_EMPTY_PASSWORD` is very confusing to me, as it seems to only have an effect on the root password. I actually tested not protecting against empty passwords and the call to `mongodb_execute` would fail. Debugging with `set -x` revealed that the mongoDB shell raised an exception with the message  `Error: couldn't add user: User passwords must not be empty`. On `mongodb_execute` failure, the script will abruptly end and the container will stop.

Do you have any information on the meaning of the history behind `ALLOW_EMPTY_PASSWORD`? Maybe should we revise the code in `mongodb_validate` as it does not make sense to allow them within `mongodb_validate` if  `ALLOW_EMPTY_PASSWORD` is true, but not performing the call to effectively create the user when the password actually is empty later on during the initialisation phase. (but this feels like a different bug?).

efrecon commented:
I pushed a few changes, I think they solve my comment above. The comment itself was caused by a subtle bug in the original implementation, a bug that had migrated to my refactoring. In the original code, authorization was turned on as soon as [either an extra user was created or a password was given to the root user](https://github.com/bitnami/bitnami-docker-mongodb/blob/b67aeb0fb84de0071f163643f3887a3b8b860bc6/4.4/debian-10/rootfs/opt/bitnami/scripts/libmongodb.sh#L449), however there was no check on `MONGODB_ROOT_PASSWORD` in the [generic case of the validation](https://github.com/bitnami/bitnami-docker-mongodb/blob/b67aeb0fb84de0071f163643f3887a3b8b860bc6/4.4/debian-10/rootfs/opt/bitnami/scripts/libmongodb.sh#L87).

juan131 reviewed (approved):
Thanks so much for adding the new warnings! They will be very useful so users can understand what's going on when a specific user is not created or what database to use for connections.

The `ALLOW_EMPTY_PASSWORD` thing is kind of a standard on Bitnami containers. You can find it on almost every Bitnami container and it's just a way to allow users running solutions such as databases without authentication. That said, depending on the solution we can find weird scenarios when it can be confusing such as the one you described here with MongoDB refusing to create users without password after authentication is enabled for other users (including the 'root' one).

Your changes look good to me!! I think we're ready to go and start with the merging process but I'd like a second opinion to ensure everything is fine, so I'll request my colleague @marcosbc to take a look.

marcosbc commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -82,13 +127,58 @@ Available options are 'primary/secondary/arbiter/hidden'"
 
+    # Deprecation warnings
+    if [[ -n "${MONGODB_USERNAME}" ]]; then
+        warn "The use of MONGODB_USERNAME is deprecated, use MONGODB_USERNAMES instead. Also, note that ',' and ';' characters are now reserved for separating fields."
```
```suggestion
        warn "MONGODB_USERNAME has been deprecated in favor of MONGODB_USERNAMES. Multiple users can be specified by separating fields via ',' or ';'."
```

marcosbc commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -82,13 +127,58 @@ Available options are 'primary/secondary/arbiter/hidden'"
+        warn "The use of MONGODB_PASSWORD is deprecated, use MONGODB_PASSWORDS instead. Also, note that ',' and ';' characters are now reserved for separating fields."
+    fi
+    if [[ -n "${MONGODB_DATABASE}" ]]; then
+        warn "The use of MONGODB_DATABASE is deprecated, use MONGODB_DATABASES instead. Also, note that ',' and ';' characters are now reserved for separating fields."
```
```suggestion
        warn "MONGODB_DATABASE has been deprecated in favor of MONGODB_DATABASES. Multiple databases can be specified by separating fields via ',' or ';'."
```

marcosbc commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -82,13 +127,58 @@ Available options are 'primary/secondary/arbiter/hidden'"
         print_validation_error "$error_message"
     fi
 
+    # Warn for users with empty passwords, as these won't be created. Maybe
+    # should we just end with an error here instead?
```
Is this true? Didn't we create the user before even with empty password?

efrecon replied:
Prior logic was confusing but it would not create the user, no. This is the `[[ -n "$MONGODB_PASSWORD" ]]` [here](https://github.com/bitnami/bitnami-docker-mongodb/blob/8b9a24334c794f2150cff9717b52ee87875c0920/3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh#L514). If you try creating a user with a blank password, MongoDB will return an exception and the initialisation will end before it has managed to perform initialisation of the container.

I believe this was a subtle bug in the original implementation. [This line](https://github.com/bitnami/bitnami-docker-mongodb/blob/8b9a24334c794f2150cff9717b52ee87875c0920/3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh#L449) would have decide when to turn on authentication, and it would be true if you had given a password for `root`, but not for the user. So you would have ended up with a `root` user with the proper password, but no "extra" user as of the test that I pointed out above.

marcosbc replied:
I see, thanks for the details! For now, let's leave it as-is (meaning no need to do any changes here).

marcosbc commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -82,13 +127,58 @@ Available options are 'primary/secondary/arbiter/hidden'"
+        warn "The use of MONGODB_USERNAME is deprecated, use MONGODB_USERNAMES instead. Also, note that ',' and ';' characters are now reserved for separating fields."
+    fi
+    if [[ -n "${MONGODB_PASSWORD}" ]]; then
+        warn "The use of MONGODB_PASSWORD is deprecated, use MONGODB_PASSWORDS instead. Also, note that ',' and ';' characters are now reserved for separating fields."
```
```suggestion
        warn "MONGODB_PASSWORD has been deprecated in favor of MONGODB_PASSWORDS. Multiple passwords can be specified by separating fields via ',' or ';'."
```

marcosbc commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -13,6 +13,50 @@
+# Returns:
+#   None
+#########################
+mongodb_auth() {
```
Instead of using the function, we could simplify the logic (check below suggestion) and include the logic in `mongodb-env.sh`. If possible, we could even remove the logic to set empty passwords.

efrecon replied:
To be frank... I don't know what is best. We had an `EXTRA` in the variable name in the original code, albeit in a slightly different form.  @juan131 had good arguments against, maybe he can chime in again. We moved away from it to achieve something that is more generic.

Moving to using comma and semi-colons:
+ Does not impact existing deployments, as authentication initialisation is only performed on new deployments. For new deployments, this would be documented as of this PR.
+ Aligns with other Bitnami's images ([kafka](https://github.com/bitnami/bitnami-docker-kafka#configuration) for example uses environment variables with an `S` at the end, advertises the use of comma as a separator, but actually [use](https://github.com/bitnami/bitnami-docker-kafka/blob/2adccca9dfbe175f9f6a978ae08dfe5dd82d3d86/2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh#L254) comma, semi-colons and whitespace in practice.)

marcosbc replied:
I'm mainly concerned with removing characters from the password. If we don't limit existing use cases, it would be great, which is why I mentioned leaving `MONGODB_PASSWORD/...` as-is and then adding `MONGODB_EXTRA_...` for extra features (with clear limitations like not allowing `,` or `;` in passwords).

I'm not sure that the Kafka approach is the ideal though, maybe @juan131 can add his thoughts in case they are different from mine.

juan131 replied:
Well.. I can see your point @marcosbc. Maintaining backwards compatibility should be our top priority here. So, although I find it very repetitive, I agree that leaving `MONGODB_PASSWORD/...` as-is and then adding `MONGODB_EXTRA_...` for extra features is going to be less problematic.

marcosbc reviewed (changes requested):
Great PR! I have added some comments, mainly to improve warning messages and some code standarizations.

However, I have a concern with respect to passwords. Note that `,` and `;` are quite common characters for passwords, and removing them would be very problematic for new deployments.

Instead of replacing `MONGODB_USER/MONGODB_PASSWORD/MONGODB_DATABASE` with `MONGODB_USERNAMES/MONGODB_PASSWORDS/MONGODB_DATABASES`, why don't we instead add some extra environment variables prefixed with `_EXTRA` (e.g. `MONGODB_EXTRA_USERNAMES`) to create those after the main initialization logic, but inside `mongodb_initialize`?

That way, it would not break deployments which wish to use `,` or `;` (or whitespace with my code proposals) in a password, as long as they use `MONGODB_PASSWORD`. However if they wanted to add multiple extra users, they would not be able to have such extra characters.

WDYT?

efrecon commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/mongodb-env.sh:
```diff
@@ -110,11 +113,15 @@ export MONGODB_DEFAULT_ENABLE_IPV6="false"
 export MONGODB_USERNAME="${MONGODB_USERNAME:-}"
 export MONGODB_PASSWORD="${MONGODB_PASSWORD:-}"
 export MONGODB_DATABASE="${MONGODB_DATABASE:-}"
+export MONGODB_USERNAMES="${MONGODB_USERNAMES:-${MONGODB_USERNAME:-}}"
+export MONGODB_PASSWORDS="${MONGODB_PASSWORDS:-${MONGODB_PASSWORD:-}}"
+export MONGODB_DATABASES="${MONGODB_DATABASES:-${MONGODB_DATABASE:-}}"
```
Good catch!

efrecon commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -13,6 +13,50 @@
 . /opt/bitnami/scripts/libfs.sh
 . /opt/bitnami/scripts/libnet.sh
 
+
+########################
+# Return field separator to use in lists. One of comma or semi-colon, comma
+# being preferred.
+# Globals:
+#   None
+# Arguments:
+#   A (list) of fields
+# Returns:
+#   The separator used within that list
+#########################
+mongodb_field_separator() {
+    if printf %s\\n "$1" | grep -q ','; then
+        echo ','
+    elif printf %s\\n "$1" | grep -q ';'; then
+        echo ';'
+    fi
+}
+
+
+########################
+# Initialise the arrays databases, usernames and passwords to contain the
+# fields from their respective environment variables. 
+# Globals:
+#   MONGODB_DATABASES, MONGODB_USERNAMES, MONGODB_PASSWORDS
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+mongodb_auth() {
+    IFS="$(mongodb_field_separator "${MONGODB_DATABASES}")" read -r -a databases <<< "${MONGODB_DATABASES}"
+    IFS="$(mongodb_field_separator "${MONGODB_USERNAMES}")" read -r -a usernames <<< "${MONGODB_USERNAMES}"
+    IFS="$(mongodb_field_separator "${MONGODB_PASSWORDS}")" read -r -a passwords <<< "${MONGODB_PASSWORDS}"
```
This is because MongoDB actually accepts passwords with a whitespace. So implementing as you propose removes yet another character from the ones that can be used in passwords.

marcosbc replied:
I guess you are right, let's see what @juan131 thinks.

juan131 replied:
I think that we should use whatever approach is less restrictive with the characters that can be used in the passwords.

efrecon commented:
I have now implemented this feature through recognising `MONGODB_EXTRA_*` variables. My latest commit has also caught up with the latest changes in master as these branches had started to diverge with the introduction of version 5.0. I am now hoping this fits the bill and the varying requirements that we have had on the initial feature.

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -489,6 +606,50 @@ mongodb_set_replicasetmode_conf() {
 }
 
+########################
+# Create the appropriate users
```
This is probably more accurate:
```suggestion
# Create a MongoDB user and provide read/write permissions on a database
```

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -489,6 +606,50 @@ mongodb_set_replicasetmode_conf() {
+# Globals:
+#   MONGODB_ROOT_PASSWORD
+# Arguments:
+#   $1 Name of user
+#   $2 Password for user
+#   $3 Name of database (empty for default database)
```
Conventions
```suggestion
#   $1 - Name of user
#   $2 - Password for user
#   $3 - Name of database (empty for default database)
```

efrecon replied:
Will fix. Conventions are good!

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -489,6 +606,50 @@ mongodb_set_replicasetmode_conf() {
+#   None
+#########################
+mongodb_create_user() {
+    local result
+
+    if [[ -n "${3:-}" ]]; then
+        # MongoDB refuses to create users without passwords. It would
+        # return the following otherwise:
+        # Error: couldn't add user: User passwords must not be empty
+        if [[ -n "${2:-}" ]]; then
+            info "Creating user '$1' in database '$3'..."
+
+            result=$(
+                mongodb_execute 'root' "$MONGODB_ROOT_PASSWORD" "" "127.0.0.1" <<EOF
+db.getSiblingDB('$3').createUser({ user: '$1', pwd: '$2', roles: [{role: 'readWrite', db: '$3'}] })
+EOF
+            )
+        else
+            warn "Cannot create user $1, no password provided"
+        fi
+    else
+        if [[ -n "${2:-}" ]]; then
+            info "Creating user '$1' in default database..."
+
+            result=$(
+                mongodb_execute 'root' "$MONGODB_ROOT_PASSWORD" "" "127.0.0.1" <<EOF
+db.getSiblingDB(db.stats().db).createUser({ user: '$1', pwd: '$2', roles: [{role: 'readWrite', db: db.getSiblingDB(db.stats().db).stats().db }] })
+EOF
+            )
+        else
+            warn "Cannot create user $1, no password provided"
+        fi
+    fi
```
I think we can simplify this block:
```suggestion
    local -r user="${1:?user is required}"
    local -r password="${2:-}"
    local -r database"${3:-}"
    local query

    if [[ -z "$password" ]]; then
        warn "Cannot create user '$user', no password provided"
        return 0
    fi
    query="db.getSiblingDB('$database').createUser({ user: '$user', pwd: '$password', roles: [{role: 'readWrite', db: '$database'}] })"
    [[ -z "$database" ]] && query="db.getSiblingDB(db.stats().db).createUser({ user: '$user'', pwd: '$password'', roles: [{role: 'readWrite', db: db.getSiblingDB(db.stats().db).stats().db }] })"
    info "Creating user '$user'..."
    mongodb_execute 'root' "$MONGODB_ROOT_PASSWORD" "" "127.0.0.1" <<< "$query"
```

efrecon replied:
Yes you are almost right!

+ I fixed the quick-edit errors on declaring `database` and the double single quotes in the query
+ Most importantly, I kept catching the result of the query in a local `result` to avoid cluttering the logs with the output of Mongo CLI

Changes upcoming

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -511,14 +672,37 @@ EOF
-        result=$(
-            mongodb_execute 'root' "$MONGODB_ROOT_PASSWORD" "" "127.0.0.1" <<EOF
-db.getSiblingDB('$MONGODB_DATABASE').createUser({ user: '$MONGODB_USERNAME', pwd: '$MONGODB_PASSWORD', roles: [{role: 'readWrite', db: '$MONGODB_DATABASE'}] })
-EOF
-        )
+        # Fill in arrays called databases, usernames and passwords with
+        # information from matching environment variables, there will only be
+        # one element at max in them.
+        mongodb_auth single
+        if [[ -n "$MONGODB_DATABASE" ]]; then
+            mongodb_create_user "${usernames[0]}" "${passwords[0]}" "${databases[0]}"
+        else
+            mongodb_create_user "${usernames[0]}" "${passwords[0]}"
+        fi
```
You already checked that `MONGODB_USERNAME ` is set, and you have check in `mongodb_create_user` to check if the password or database name is set. Then I guess you can simply do this:
```suggestion
        mongodb_create_user "$MONGODB_USERNAME" "$MONGODB_PASSWORD" "$MONGODB_DATABASE"
```

efrecon replied:
Good one!

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -1272,41 +1459,3 @@ mongodb_execute() {
     "$MONGODB_BIN_DIR/mongo" "${args[@]}"
 }
-
-########################
-# Execute an arbitrary query/queries against the running MongoDB service
-# Stdin:
-#   Query/queries to execute
-# Arguments:
-#   $1 - User to run queries
-#   $2 - Password
-#   $3 - Database where to run the queries
-#   $4 - Host (default to result of get_mongo_hostname function)
-#   $5 - Port (default $MONGODB_PORT_NUMBER)
-#   $6 - Extra arguments (default $MONGODB_CLIENT_EXTRA_FLAGS)
-# Returns:
-#   None
-########################
-mongodb_execute() {
-    local -r user="${1:-}"
-    local -r password="${2:-}"
-    local -r database="${3:-}"
-    local -r host="${4:-$(get_mongo_hostname)}"
-    local -r port="${5:-$MONGODB_PORT_NUMBER}"
-    local -r extra_args="${6:-$MONGODB_CLIENT_EXTRA_FLAGS}"
-    local final_user="$user"
-    # If password is empty it means no auth, do not specify user
-    [[ -z "$password" ]] && final_user=""
-
-    local -a args=("--host" "$host" "--port" "$port")
-    [[ -n "$final_user" ]] && args+=("-u" "$final_user")
-    [[ -n "$password" ]] && args+=("-p" "$password")
-    if [[ -n "$extra_args" ]]; then
-        local extra_args_array=()
-        read -r -a extra_args_array <<<"$extra_args"
-        [[ "${#extra_args_array[@]}" -gt 0 ]] && args+=("${extra_args_array[@]}")
-    fi
-    [[ -n "$database" ]] && args+=("$database")
-
-    "$MONGODB_BIN_DIR/mongo" "${args[@]}"
-}
```
Why did you remove this?

efrecon replied:
because... it was declared twice!

juan131 reviewed (changes requested):
Wow!! This is amazing!! Thanks so much @efrecon 

Please check my comments when you have a chance.

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -607,47 +607,33 @@ mongodb_set_replicasetmode_conf() {
+    [[ -z "$database" ]] && query="db.getSiblingDB(db.stats().db).createUser({ user: '$user', pwd: '$password', roles: [{role: 'readWrite', db: db.getSiblingDB(db.stats().db).stats().db }] })"
+    # Create user, discarding mongo CLI output for clean logs
+    info "Creating user '$user'..."
+    result=$(mongodb_execute 'root' "$MONGODB_ROOT_PASSWORD" "" "127.0.0.1" <<< "$query")
```
You don't need to capture the result in a variable. Luckily we have a function for that: `debug_execute`. See:

- https://github.com/bitnami/bitnami-docker-mongodb/blob/master/5.0/debian-10/prebuildfs/opt/bitnami/scripts/libos.sh#L252

This function will redirect the output to `/dev/null` unless the `BITNAMI_DEBUG` variable is set to `true`. I propose keeping this line like this:

```bash
    mongodb_execute 'root' "$MONGODB_ROOT_PASSWORD" "" "127.0.0.1" <<< "$query"

```

And.. Modifying the `mongodb_execute` like this:

```diff
-    "$MONGODB_BIN_DIR/mongo" "${args[@]}"
+    debug_execute "$MONGODB_BIN_DIR/mongo" "${args[@]}"
```

There are parts of the library where we are really interested in the command output (regardless the value set in the `BITNAMI_DEBUG` env. var). We can force `BITNAMI_DEBUG=false` in these cases:

```bash
result=$(BITNAMI_DEBUG=false mongodb_execute ...)
```

juan131 reviewed (commented):
A small comment to manage logs properly! The rest LGTM!!
@marcosbc could you take also a look?

juan131 commented:
Please also rebase your branch and fix the conflicts when you have a chance.

efrecon commented:
I considered doing exactly what you proposed, but then I figured out that most uses of `mongodb_execute` made actual use of the result in their following logic. So, instead, I created a `mongodb_concealed_execute` that does exactly what the function name tells: it conceals the output of `mongodb_execute` unless `BITNAMI_DEBUG` is `true`. I think this was for the best.

juan131 commented on 3.6/debian-9/rootfs/opt/bitnami/scripts/libmongodb.sh:
```diff
@@ -1287,26 +1448,6 @@ mongodb_execute() {
-}
+mongodb_concealed_execute() {
+    debug_execute mongodb_execute "$@"
+}
```
Please fix the "No newline at end of file" warning.

juan131 reviewed (commented):
I love the idea of having two functions: a "silent" one and one that prints the output. Thanks!
Just for being consistent with other Bitnami libraries, can we name then `mongodb_execute_print_output` and `mongodb_execute`? Check MySQL:
- https://github.com/bitnami/bitnami-docker-mysql/blob/master/8.0/debian-10/rootfs/opt/bitnami/scripts/libmysql.sh#L495
- https://github.com/bitnami/bitnami-docker-mysql/blob/master/8.0/debian-10/rootfs/opt/bitnami/scripts/libmysql.sh#L534

efrecon commented:
So @juan131, this should use the names that you requested (and has an extra newline!)

juan131 reviewed (approved):
LGTM

juan131 commented:
Thanks so much @efrecon !! I really appreciate the great work you did!! 👏🏼 
@marcosbc could you please also review these changes again?

marcosbc reviewed (changes requested):
Thanks a lot for these amazing changes! I have some small requests, could you have a look into them?
